T6291: Add LACP related op commands

l0crian1 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
This PR adds additional op commands for LACP related information. The following commands will be added:

`show interfaces bonding lacp detail`

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
https://vyos.dev/T6291

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
/op-mode-definitions/show-interfaces-bonding.xml.in
New File: /src/op_mode/show_lacp.py

## Proposed changes
<!--- Describe your changes in detail -->
This change will add the following output for LACP related information:

```
vyos@vyos:~$ show interfaces bonding lacp detail 
Interface    Members    Mode    Rate    System MAC         Hash
-----------  ---------  ------  ------  -----------------  ------
bond0        eth2       active  slow    16:a4:1f:8b:c0:b3  layer2
bond1                   active  slow    00:00:00:00:00:00  layer2
```

```
vyos@vyos:~$ show interfaces bonding bond0 lacp detail 
Interface    Members    Mode    Rate    System MAC         Hash
-----------  ---------  ------  ------  -----------------  ------
bond0        eth2       active  slow    16:a4:1f:8b:c0:b3  layer2
```

```
vyos@vyos:~$ show interfaces bonding bond0 lacp neighbors 
Interface    Member    Local ID           Remote ID
-----------  --------  -----------------  -----------------
bond0        eth2      16:a4:1f:8b:c0:b3  00:00:00:00:00:00
```

Additionally, the `show interfaces bonding <bondif> detail` command is updated to be run as sudo, allowing for output of valuable information.
## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
Configure bond interfaces in mode 802.3ad and run new commands:
```
vyos@vyos# set interfaces bonding bond0 mode 802.3ad 
vyos@vyos# commit

vyos@vyos:~$ show interfaces bonding lacp detail 
Interface    Members    Mode    Rate    System MAC         Hash
-----------  ---------  ------  ------  -----------------  ------
bond0        eth2       active  slow    16:a4:1f:8b:c0:b3  layer2
bond1                   active  slow    00:00:00:00:00:00  layer2
```

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
new file:   src/op_mode/show_lacp.py

T6291: Add additional op commands for LACP related info. New Command:
show interfaces bonding lacp detail
show interfaces bonding <bondif> lacp detail
show interfaces bonding <bondif> lacp neighbors

l0crian1 committed:
modified:   op-mode-definitions/show-interfaces-bonding.xml.in
new file:   src/op_mode/show_lacp.py

T6291: Add addtional op commands for LACP related info. New Command:
show interfaces bonding lacp detail
show interfaces bonding <bondif> lacp detail
show interfaces bonding <bondif> lacp neighbors

l0crian1 committed:
T6291: modified:   src/op_mode/show_lacp.py

Updated show_lacp.py comment section
Corrected error where script failed for
"show interfaces bonding lacp detal" when no bonds were configured

l0crian1 committed:
T6291: modified:   op-mode-definitions/show-interfaces-bonding.xml.in

Fixed error in script call

l0crian1 committed:
T6291: modified:   src/op_mode/show_lacp.py
Fixed show_lacp.py permissions

l0crian1 committed:
T6291: modified:   op-mode-definitions/show-interfaces-bonding.xml.in

Updated "show interfaces bonding detail" command to be ran as sudo
for improved output

dmbaturin commented on src/op_mode/show_lacp.py:
```diff
@@ -0,0 +1,101 @@
+    else:
+        print("No member interfaces found!")
+
+if __name__ == "__main__":
```
All new operational mode scripts must use the new approavh with `vyos.opmode` so that they can be automatically exposed in the API as well.
See https://blog.vyos.io/writing-vyos-operational-mode-commands-just-got-better and a new-style script like https://github.com/vyos/vyos-1x/blob/current/src/op_mode/cpu.py

l0crian1 replied:
Is there a specific format the dictionary output needs to be?

l0crian1 replied:
Is something like this sufficient:
```
{
    "lacp": [
        {
            "interface": "bond0",
            "members": "eth3",
            "mode": "active",
            "rate": "slow",
            "system_mac": "aa:bb:cc:dd:ee:f0",
            "hash": "layer2"
        },
        {
            "interface": "bond1",
            "members": "",
            "mode": "active",
            "rate": "slow",
            "system_mac": "aa:bb:cc:dd:ee:f0",
            "hash": "layer2"
        }
    ]
}
```

l0crian1 replied:
@dmbaturin I updated the script to use vyos.opmode

l0crian1 committed:
T6291: modified:   op-mode-definitions/show-interfaces-bonding.xml.in
T6291: deleted:    src/op_mode/show_lacp.py
Updated show_lacp.py script to new vyos.opmode scheme
Renamed show_lacp.py to bonding.py

l0crian1 committed:
T6291: Changed show_lacp.py to bonding.py
T6291: Updated bonding.py to use vyos.opmode scheme

l0crian1 committed:
T6291: Renamed show_lacp.py to bonding.py
T6291: Updated bonding.py to use vyos.opmode

l0crian1 committed:
T6291: Updated command section in show-interfaces-bonding.xml.in

sever-sever commented on src/op_mode/bonding.py:
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2016-2022 VyOS maintainers and contributors
```
2024 ;)

l0crian1 replied:
Thanks, good catch! Updated!

l0crian1 committed:
T6291: Updated copyright year in bonding.py

The previous copyright info had 2016-2022; updated to 2016-2024

l0crian1 commented:
Looks like I need to update a portion of the script related to error handling as well. Is this the correct updates for the vyos.opmode approach?:
```
    if 'Bonding Mode: IEEE 802.3ad Dynamic link aggregation' not in data:
        raise vyos.opmode.DataUnavailable("Bond is not present or not configured with mode 802.3ad")
```
And update the exception in the main block:
```
if __name__ == '__main__':
    try:
        res = vyos.opmode.run(sys.modules[__name__])
        if res:
            print(res)
    except (ValueError, vyos.opmode.Error) as e:
        print(e)
        sys.exit(1)
```

jestabro commented:
You may want to squash these commits, when changes are complete. Also, one common construction can avoid the call to 'cli-shell-api':

```
diff --git a/src/op_mode/bonding.py b/src/op_mode/bonding.py
index 9bd69d2b7..f5e289db3 100755
--- a/src/op_mode/bonding.py
+++ b/src/op_mode/bonding.py
@@ -16,12 +16,14 @@
 #
 # This script will parse 'sudo cat /proc/net/bonding/<interface name>' and return table output for lacp related info

-from tabulate import tabulate
 import subprocess
 import re
 import sys
 import typing
+from tabulate import tabulate
+
 import vyos.opmode
+from vyos.configquery import ConfigTreeQuery

 def list_to_dict(data, headers, basekey):
     data_list = {basekey: []}
@@ -61,11 +63,12 @@ def show_lacp_neighbors(raw: bool, interface: typing.Optional[str]):

 def show_lacp_detail(raw: bool, interface: typing.Optional[str]):
     headers = ["Interface", "Members", "Mode", "Rate", "System-MAC", "Hash"]
+    query = ConfigTreeQuery()

     if interface:
         intList = [interface]
     else:
-        intList = subprocess.run(f"cli-shell-api listActiveNodes interfaces bonding", stdout=subprocess.PIPE, shell=True, text=True).stdout.encode('utf-8').decode('utf-8').replace("'", "").split()
+        intList = query.list_nodes(['interfaces', 'bonding'])

     bondList = []

-- 
2.44.0
```

l0crian1 commented:
Thanks @jestabro that's definitely cleaner. I'm not sure how to squash the commits, but I'll try once everybody is happy with the PR.

l0crian1 committed:
T6291: Updated bonding.py to exit with proper code

I previously was using the exit() function to exit the script under certain conditions. This was not in line with the vyos.opmode approach to op scripts. The bonding.py script was updated to leverage the DataUnavailable() function to provide the appropriate error code if using the API.

c-po commented on src/op_mode/bonding.py:
```diff
@@ -0,0 +1,103 @@
+
+def show_lacp_neighbors(raw: bool, interface: typing.Optional[str]):
+    headers = ["Interface", "Member", "Local ID", "Remote ID"]
+    data = subprocess.run(f"sudo cat /proc/net/bonding/{interface}", stdout=subprocess.PIPE, stderr=subprocess.DEVNULL, shell=True, text=False).stdout.decode('utf-8')
```
`sudo` call should be moved outside the script, to the caller in the XML definition.

c-po commented on src/op_mode/bonding.py:
```diff
@@ -0,0 +1,103 @@
+    bondList = []
+
+    for interface in intList:
+        data = subprocess.run(f"sudo cat /proc/net/bonding/{interface}", stdout=subprocess.PIPE, stderr=subprocess.DEVNULL, shell=True, text=False).stdout.decode('utf-8')
```
`sudo` call should be moved outside the script, to the caller in the XML definition.

l0crian1 replied:
Done!

c-po commented:
Other then the two `sudo` calls the rest looks okay and we can squash it afterwards

l0crian1 committed:
T6291: Removed sudo from subprocess calls

Removed sudo from subprocess calls. The sudo elevation will be handled in the op-mode-definition XML.

l0crian1 committed:
T6291: Moved sudo calls from bonding.py to op definition

Moved sudo calls within the bonding.py script into the op XML definition for elevation.

c-po commented:
@MergifyIo backport sagitta
